cpp/src/arrow/flight/client.cc

```diff
@@ -259,6 +111,65 @@ class FlightStreamReader : public RecordBatchReader {
   std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream_;
 };
 
+class FlightClient;
+
+/// \brief A RecordBatchWriter implementation that writes to a Flight
+/// DoPut stream.
+class FlightStreamWriter : public ipc::RecordBatchWriter {
+ public:
+  explicit FlightStreamWriter(std::unique_ptr<ClientRpc>&& rpc,
+                              const FlightDescriptor& descriptor,
+                              const std::shared_ptr<Schema>& schema)
```

Comment:
Pass in `MemoryPool* pool = default_memory_pool()` maybe (though you can use `set_memory_pool`)?